tests/generator/test_veths.py

```diff
@@ -0,0 +1,222 @@
+  renderer: NetworkManager
+  veths:
+    veth0:
+      peer: veth1''', confs={'b': '''network:
+  version: 2
+  renderer: NetworkManager
+  veths:
+    veth1:
+      peer: asd'''}, expect_fail=True)
+
+        self.assertIn('veth peer \'veth1\' does not have a peer itself', out)
+
+    def test_veth_peer_is_anothers_interface_peer_already(self):
+        out = self.generate('''network:
+  version: 2
+  renderer: NetworkManager
+  veths:
+    veth2:
+      peer: veth1
+    veth0:
+      peer: veth1
+    veth1:
+      peer: veth2''', expect_fail=True)
+
+        self.assertIn('veth peer \'veth1\' is another veth\'s (veth2) peer already', out)
+
+    def test_veth_peer_is_anothers_interface_peer_already_validation_stage(self):
+        out = self.generate('''network:
+  version: 2
+  renderer: NetworkManager
+  veths:
+    veth0:
+      peer: veth1
+      ''', confs={'b': '''network:
+  renderer: NetworkManager
+  veths:
+    veth1:
+      peer: veth2
+    veth2:
+      peer: veth1'''}, expect_fail=True)
+
+        self.assertIn('veth peer \'veth1\' is another veth\'s (veth2) peer already', out)
+
+    def test_veth_peer_is_itself(self):
+        out = self.generate('''network:
+  version: 2
+  renderer: NetworkManager
+  veths:
+    veth1:
+      peer: veth1''', expect_fail=True)
+
+        self.assertIn('veth peer cannot be itself', out)
+
+
+class TestNetworkd(TestBase):
+
+    def test_basic(self):
+        self.generate('''network:
+  version: 2
+  veths:
+    veth0:
+      peer: veth1
+    veth1:
+      peer: veth0''')
+
+        self.assert_networkd({'veth0.network': ND_EMPTY % ('veth0', 'ipv6'),
+                              'veth1.network': ND_EMPTY % ('veth1', 'ipv6'),
+                              'veth0.netdev': ND_VETH % ('veth0', 'veth1')})
+
+    def test_basic_missing_peer(self):
+        ''' When networkd is the renderer, both peers are required '''
+        out = self.generate('''network:
+  version: 2
+  veths:
+    veth0:
+      peer: veth1''', expect_fail=True)
+
+        self.assertIn('veth0: interface \'veth1\' is not defined', out)
+
+    def test_missing_peer_key_should_fail(self):
+        out = self.generate('''network:
+  version: 2
+  renderer: networkd
+  veths:
+    veth0: {}''', expect_fail=True)
+
+        self.assertIn('veth missing \'peer\' property', out)
+
+    def test_veth_peer_is_not_a_veth_interface(self):
+        out = self.generate('''network:
+  version: 2
+  renderer: networkd
+  ethernets:
+    eth0: {}
+  veths:
+    veth0:
+      peer: eth0''', expect_fail=True)
+
+        self.assertIn('\'eth0\' is not a veth interface', out)
+
+    def test_veth_peer_of_a_peer_was_not_defined(self):
+        out = self.generate('''network:
+  version: 2
+  renderer: networkd
+  veths:
+    veth0:
+      peer: veth1
+    veth1:
+      peer: abc''', expect_fail=True)
+
+        self.assertIn('veth1: interface \'abc\' is not defined', out)
+
+    def test_veth_peer_is_anothers_interface_peer_already(self):
```

Comment:
nitpick (non-blocking): I guess we could combine some of those tests into a `_CommonTests()` class (where the behavior does not differ between the backend renderers), to make equal tests run on both backends. Cf. `tests/generator/test_tunnels.py`